c_src/sqlite3_nif.c

```diff
@@ -1157,29 +1221,36 @@ exqlite_interrupt(ErlNifEnv* env, int argc, const ERL_NIF_TERM argv[])
     return make_atom(env, "ok");
 }
 
-//
-// Most of our nif functions are going to be IO bounded
-//
-
 static ErlNifFunc nif_funcs[] = {
-  {"open", 2, exqlite_open, ERL_NIF_DIRTY_JOB_IO_BOUND},
-  {"close", 1, exqlite_close, ERL_NIF_DIRTY_JOB_IO_BOUND},
-  {"execute", 2, exqlite_execute, ERL_NIF_DIRTY_JOB_IO_BOUND},
-  {"changes", 1, exqlite_changes, ERL_NIF_DIRTY_JOB_IO_BOUND},
-  {"prepare", 2, exqlite_prepare, ERL_NIF_DIRTY_JOB_IO_BOUND},
-  {"bind", 3, exqlite_bind, ERL_NIF_DIRTY_JOB_IO_BOUND},
-  {"step", 2, exqlite_step, ERL_NIF_DIRTY_JOB_IO_BOUND},
-  {"multi_step", 3, exqlite_multi_step, ERL_NIF_DIRTY_JOB_IO_BOUND},
-  {"columns", 2, exqlite_columns, ERL_NIF_DIRTY_JOB_IO_BOUND},
-  {"last_insert_rowid", 1, exqlite_last_insert_rowid, ERL_NIF_DIRTY_JOB_IO_BOUND},
-  {"transaction_status", 1, exqlite_transaction_status, ERL_NIF_DIRTY_JOB_IO_BOUND},
-  {"serialize", 2, exqlite_serialize, ERL_NIF_DIRTY_JOB_IO_BOUND},
-  {"deserialize", 3, exqlite_deserialize, ERL_NIF_DIRTY_JOB_IO_BOUND},
-  {"release", 2, exqlite_release, ERL_NIF_DIRTY_JOB_IO_BOUND},
-  {"enable_load_extension", 2, exqlite_enable_load_extension, ERL_NIF_DIRTY_JOB_IO_BOUND},
-  {"set_update_hook", 2, exqlite_set_update_hook, ERL_NIF_DIRTY_JOB_IO_BOUND},
-  {"set_log_hook", 1, exqlite_set_log_hook, ERL_NIF_DIRTY_JOB_IO_BOUND},
-  {"interrupt", 1, exqlite_interrupt, ERL_NIF_DIRTY_JOB_IO_BOUND},
+  {"dirty_io_open", 2, exqlite_open, ERL_NIF_DIRTY_JOB_IO_BOUND},
+  {"dirty_io_close", 1, exqlite_close, ERL_NIF_DIRTY_JOB_IO_BOUND},
+  {"dirty_io_execute", 2, exqlite_execute, ERL_NIF_DIRTY_JOB_IO_BOUND},
+  {"dirty_io_step", 2, exqlite_step, ERL_NIF_DIRTY_JOB_IO_BOUND},
+  {"dirty_io_serialize", 2, exqlite_serialize, ERL_NIF_DIRTY_JOB_IO_BOUND},
+  {"dirty_io_deserialize", 3, exqlite_deserialize, ERL_NIF_DIRTY_JOB_IO_BOUND},
+  {"dirty_io_interrupt", 1, exqlite_interrupt, ERL_NIF_DIRTY_JOB_IO_BOUND},
+
+  {"dirty_io_multi_step", 3, exqlite_multi_step, ERL_NIF_DIRTY_JOB_IO_BOUND},
+  {"dirty_io_insert_all", 3, exqlite_insert_all, ERL_NIF_DIRTY_JOB_IO_BOUND},
+
+  {"dirty_cpu_prepare", 2, exqlite_prepare, ERL_NIF_DIRTY_JOB_CPU_BOUND},
+  {"dirty_cpu_bind_all", 3, exqlite_bind_all, ERL_NIF_DIRTY_JOB_CPU_BOUND},
+
+  {"execute", 2, exqlite_execute},
+  {"changes", 1, exqlite_changes},
+  {"prepare", 2, exqlite_prepare},
+  {"columns", 2, exqlite_columns},
+  {"step", 2, exqlite_step},
+  {"interrupt", 1, exqlite_interrupt},
+  {"finalize", 1, exqlite_finalize},
+  {"last_insert_rowid", 1, exqlite_last_insert_rowid},
+  {"transaction_status", 1, exqlite_transaction_status},
```

Comment:
Why are these not going through the Dirty NIF? These are io bound, specifically `exqlite_step` is definitely IO bound.

Reply:
👋 @warmwaffles 

Not necessarily. Sometimes `exqlite_step` can be safely scheduled on non-dirty thread, this change introduces multiple versions of each nif, making it possible for the caller to decide which one to use. By default, the library would keep using dirty schedulers but it will also introduce a family of "unsafe" implementations :)

Reply:
I decided to do this split when I noticed that non-dirty exqlite_step is almost twice as performant as a dirty exqlite_multi_step on all data sets I tried them on.

Reply:
> I'll look into simulating slow disk access (system under heavy write mode)

And I'll try to find their implementation. Right now I think they just read from internal connection state.

Reply:
> I decided to do this split when I noticed that non-dirty `exqlite_step` is almost twice as performant as a dirty `exqlite_multi_step` on all data sets I tried them on.

I originally did the multi step implementation because I had a very large sqlite database of timeseries data and it did not have a lot of throughput. But of course that was on the implementation I wrote, and what you have here _may_ actually be better.